MailAlert.php

```diff
@@ -205,7 +205,7 @@ public static function sendCustomerAlert($id_product, $id_product_attribute)
                     Mail::Send(
                         $id_lang,
                         'customer_qty',
-                        Mail::l('Product available', $id_lang),
+                        $this->trans('Product available', array(), 'Emails.Subject'),
```

Comment:
The method `trans` is not defined in this class. So you must inject either the translator or the already translated message.

Reply:
Can you explain how to do that?

Reply:
If you want to inject the translator into `MailAlert` class in order to provide it with translating capabilities, you must use [Dependency Injection](https://en.wikipedia.org/wiki/Dependency_injection).

However I think we cannot control how the `__construct()` of MailAlert is called, because it's an ObjectModel, so we might use a dedicated setter. Moreovor the function you want to modify is a static class. This means:

```
// new property of the class ; it has to be static to allow us to use it in a static context
private static $translator;

// new function to allow the injection
public static function setTranslator($translator)
{
     self::$translator = $translator;
}

public static function sendCustomerAlert(...)
{
    ...
    if (null === self::$translator) {
        // as we cannot be 100% sure the translator has been injected using setTranslator(), better check it
        throw new \RuntimeException('Mandatory translator is not available');
    }

    Mail::Send(
        $id_lang,
        'customer_qty',
        self::$translator->trans('Product available', array(), 'Emails.Subject'),
        ...
    );
}
```

AND we need to make sure to inject the translator, from the main module class:
```
// ps_emailalerts.php

class Ps_EmailAlerts extends Module
{
    ...
    public function __construct()
    {
        MailAlert::setTranslator($this->getTranslator());
    }

Reply:
Can we get the translator throw the controller->getModule or something like that ?

Reply:
Although it's not a good practice, you can get the translator from context:

```php
$translator = Context::getContext()->getTranslator();
```

Bu warning: this translator will be initialized in the current context's language, which is not necessarily the same language as the recipients's. I'm not familiar with this module so I don't know in which context this code is executed.

If you need to load the translator in a specific language, you can do so using this method (available since 1.7.3.1):

```php
$locale = 'en-US'; // IETF language tag
Context::getContext()->getTranslatorFromLocale($locale);
```